aws_lambda_builders/workflows/ruby/DESIGN.md

```diff
@@ -0,0 +1,42 @@
+# Ruby - Lambda Builder
+
+## Scope
+
+For the basic case, building the dependencies for a Ruby Lambda project is very easy:
+
+```shell
+# ensure you are using Ruby 2.5, for example with rbenv or rvm
+bundle install # if no Gemfile.lock is present
+bundle install --deployment
+zip -r source.zip * # technically handled by `sam package`
+```
+
+The basic scope of a `sam build` script for Ruby would be as a shortcut for this, while performing some housekeeping steps:
+
+- Skipping the initial `bundle install` if a Gemfile.lock file is present.
+- Ensuring that `ruby --version` matches `/^ ruby 2\.5\./`
+- Raising a soft error if there is already a `.bundle` and `vendor/bundle` folder structure, and giving an option to clobber this if desired.
+  - I don't want this to be a default behavior, in case users are using the `vendor` or `.bundle` folder structures for other things and clobbering it could have destructive and unintended side effects.
+
+Having a unified command also gives us the ability to solve once the most common issues and alternative use cases in a way that follows best practices:
+
+1. Including dependencies that have native extensions, and building them in the proper environment.
+   - An open question is how to help users represent binary dependencies, but that's not a Ruby concern per se so it should be solved the same way across all builds.
+2. Building and deploying the user dependencies as a layer rather than as part of the code package.
+   - These also have slightly different folder pathing:
+     - Bundled dependencies are looked for in `/var/task/vendor/bundle/ruby/2.5.0` which is the default result of a `bundle install --deployment` followed by an upload.
+     - Layer dependencies are looked for in `/opt/ruby/gems/2.5.0`, so for a layer option would have to use a `--path` build or transform the folder structure slightly.
+3. Down the road, perhaps providing a way to bundle code as a layer, such as for shared libraries that are not gems. These need to go in the `/opt/ruby/lib` folder structure.
+
+## Challenges
+
+- Ensuring that builds happen in Ruby 2.5.x only.
+- Ensuring that builds that include native extensions happen in the proper build environment.
+
+## Interface/Implementation
+
+Off hand, I envision the following commands as a starting point:
```

Comment:
How do you envision the workflow actions to look like?
would it just be?
Action 1: CopySource
Action 2: bundle install # if no Gemfile.lock is present
Action 3: bundle install --deployment

That seems nice and simple to start out 👍